File: papermc/src/main/java/net/eve0415/spigot/WebsocketIntegration/PaperEventListener.java
package net.eve0415.spigot.WebsocketIntegration;

import java.util.UUID;

import com.destroystokyo.paper.event.player.PlayerConnectionCloseEvent;

import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.PlayerDeathEvent;
import org.bukkit.event.player.AsyncPlayerPreLoginEvent;
import org.bukkit.event.player.PlayerAdvancementDoneEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerKickEvent;
import org.bukkit.event.player.PlayerLoginEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.event.player.PlayerLoginEvent.Result;
import org.json.JSONException;

import io.papermc.paper.event.player.AsyncChatEvent;
import net.eve0415.spigot.WebsocketIntegration.Util.LogEventType;
import net.eve0415.spigot.WebsocketIntegration.Util.WSIEventState;
import net.kyori.adventure.text.Component;

public class PaperEventListener implements Listener {
    public PaperEventListener(final WSIPaperPlugin instance) {
        instance.getServer().getPluginManager().registerEvents(this, instance);
    }

    private void sendChatMessage(final String name, final UUID uuid, final Component message) {
        try {
            WebsocketManager.getInstance().send(WSIEventState.CHAT,
                    WebsocketManager.builder().message(name, uuid, message).toJSON());
        } catch (final JSONException e) {
            WebsocketManager.getInstance().getWSILogger().error("There was an error trying to send chat message", e);
        }
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onJoin(final PlayerJoinEvent event) {
        sendChatMessage(event.getPlayer().getName(), event.getPlayer().getUniqueId(), event.joinMessage());
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onQuit(final PlayerQuitEvent event) {
        sendChatMessage(event.getPlayer().getName(), event.getPlayer().getUniqueId(), event.quitMessage());
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onDeath(final PlayerDeathEvent event) {
        sendChatMessage(event.getEntity().getName(), event.getEntity().getUniqueId(), event.deathMessage());
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onMessage(final AsyncChatEvent event) {
        sendChatMessage(event.getPlayer().getName(), event.getPlayer().getUniqueId(), event.message());
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onAdvancementDone(final PlayerAdvancementDoneEvent event) {
        // We don't want to call out receipes unlock as advancement complete.
        if (event.getAdvancement().getKey().getKey().toString().contains("recipes/"))
            return;

        final String adv = event.getAdvancement().getKey().getNamespace().toUpperCase() + "_"
                + event.getAdvancement().getKey().getKey().replace('/', '_').toUpperCase();

        try {
            WebsocketManager.getInstance().send(WSIEventState.ADVANCEMENT, WebsocketManager.builder()
                    .message(event.getPlayer().getName(), event.getPlayer().getUniqueId(), adv).toJSON());
        } catch (final JSONException e) {
            WebsocketManager.getInstance().getWSILogger().error("There was an error trying to send chat message", e);
        }
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onPlayerPreLogin(final AsyncPlayerPreLoginEvent event) {
        try {
            WebsocketManager.getInstance().send(WSIEventState.LOG, WebsocketManager.builder()
                    .log(LogEventType.AUTH, event.getName(), event.getUniqueId(), event.getAddress().getHostAddress())
                    .toJSON());
        } catch (final JSONException e) {
            e.printStackTrace();
        }
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onPlayerLogin(final PlayerLoginEvent event) {
        try {
            if (event.getResult() == Result.ALLOWED) {
                WebsocketManager.getInstance().send(WSIEventState.LOG,
                        WebsocketManager
                                .builder().log(LogEventType.LOGIN, event.getPlayer().getName(),
                                        event.getPlayer().getUniqueId(), event.getAddress().getHostAddress())
                                .setAddress(event.getHostname()).toJSON());
            } else {
                WebsocketManager.getInstance().send(WSIEventState.LOG,
                        WebsocketManager
                                .builder().log(LogEventType.KICK, event.getPlayer().getName(),
                                        event.getPlayer().getUniqueId(), event.getHostname())
                                .kick(event.kickMessage()).toJSON());
            }
        } catch (final JSONException e) {
            e.printStackTrace();
        }
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onPlayerKick(final PlayerKickEvent event) {
        try {
            WebsocketManager.getInstance().send(WSIEventState.LOG,
                    WebsocketManager
                            .builder().log(LogEventType.KICK, event.getPlayer().getName(),
                                    event.getPlayer().getUniqueId(), event.getPlayer().getAddress().getHostName())
                            .kick(event.reason()).toJSON());
        } catch (final JSONException e) {
            e.printStackTrace();
        }
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onPlayerConnectionClose(final PlayerConnectionCloseEvent event) {
        try {
            WebsocketManager.getInstance().send(WSIEventState.LOG,
                    WebsocketManager.builder().log(LogEventType.DISCONNECT, event.getPlayerName(),
                            event.getPlayerUniqueId(), event.getIpAddress().getHostName()).toJSON());
        } catch (final JSONException e) {
            e.printStackTrace();
        }
    }
}

File: paper/src/main/kotlin/net/eve0415.mc/websocketintegration/PaperEventListener.kt
package net.eve0415.mc.websocketintegration

import com.destroystokyo.paper.event.player.PlayerConnectionCloseEvent
import io.papermc.paper.event.player.AsyncChatEvent
import net.eve0415.mc.websocketintegration.type.LogEventType
import net.eve0415.mc.websocketintegration.type.WIEventState
import net.kyori.adventure.text.Component
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.entity.PlayerDeathEvent
import org.bukkit.event.player.*
import java.util.*


class PaperEventListener constructor(private val instance: PaperPlugin) : Listener {
  private val manager: WebsocketManager

  init {
    instance.server.pluginManager.registerEvents(this, instance)
    manager = instance.websocketManager
  }

  private fun sendChatMessage(name: String, uuid: UUID, message: Component) {
    manager.send(
      WIEventState.CHAT,
      manager.builder().message(name, uuid, message).toJSON()
    )
  }

  @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
  fun onJoin(event: PlayerJoinEvent) {
    event.joinMessage()?.let { sendChatMessage(event.player.name, event.player.uniqueId, it) }
  }

  @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
  fun onQuit(event: PlayerQuitEvent) {
    event.quitMessage()?.let { sendChatMessage(event.player.name, event.player.uniqueId, it) }
  }

  @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
  fun onDeath(event: PlayerDeathEvent) {
    event.deathMessage()?.let { sendChatMessage(event.entity.name, event.entity.uniqueId, it) }
  }

  @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
  fun onMessage(event: AsyncChatEvent) {
    sendChatMessage(event.player.name, event.player.uniqueId, event.message())
  }

  @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
  fun onAdvancementDone(event: PlayerAdvancementDoneEvent) {
    // We don't want to call out recipes unlock as advancement complete.
    if (event.advancement.key.key.contains("recipes/")) return
    val adv = "${event.advancement.key.namespace.uppercase()}_${
      event.advancement.key.key.replace('/', '_').uppercase()
    }"

    manager.send(
      WIEventState.ADVANCEMENT, manager.builder()
        .message(event.player.name, event.player.uniqueId, adv).toJSON()
    )
  }

  @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
  fun onPlayerPreLogin(event: AsyncPlayerPreLoginEvent) {
    manager.send(
      WIEventState.LOG, manager.builder()
        .log(LogEventType.AUTH, event.name, event.uniqueId, event.address.hostAddress)
        .toJSON()
    )
  }

  @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
  fun onPlayerLogin(event: PlayerLoginEvent) {
    if (event.result == PlayerLoginEvent.Result.ALLOWED) {
      manager.send(
        WIEventState.LOG,
        manager
          .builder().log(
            LogEventType.LOGIN, event.player.name,
            event.player.uniqueId, event.address.hostAddress
          )
          .setAddress(event.hostname).toJSON()
      )
    } else {
      manager.send(
        WIEventState.LOG,
        manager
          .builder().log(
            LogEventType.KICK, event.player.name,
            event.player.uniqueId, event.hostname
          )
          .kick(event.kickMessage()).toJSON()
      )
    }
  }

  @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
  fun onPlayerKick(event: PlayerKickEvent) {
    manager.send(
      WIEventState.LOG,
      manager
        .builder().log(
          LogEventType.KICK, event.player.name,
          event.player.uniqueId, event.player.address.hostName
        )
        .kick(event.reason()).toJSON()
    )
  }

  @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
  fun onPlayerConnectionClose(event: PlayerConnectionCloseEvent) {
    manager.send(
      WIEventState.LOG,
      manager.builder().log(
        LogEventType.DISCONNECT, event.playerName,
        event.playerUniqueId, event.ipAddress.hostName
      ).toJSON()
    )
  }
}
